Fast Fail Impossible Routes Seems Broken

Recently I tried to do this:
![image](https://user-images.githubusercontent.com/697548/124982388-a2c14c00-e004-11eb-9d5f-6f7aa865cf51.png)

This should result in an immediate 400 and a message about how the locations are in disconnected regions however i saw this:
![image](https://user-images.githubusercontent.com/697548/124982458-b8cf0c80-e004-11eb-8bad-e238523a6aca.png)

This request should never make it to thor so long as loki is using the connectivity map so i checked my config and:
![image](https://user-images.githubusercontent.com/697548/124982561-d56b4480-e004-11eb-8c3e-75520be8a550.png)

It looks like it should be!

i modified `valhalla_export_edges` adding this to the top of the inner for loop:

```c++
      auto inf = tile->edgeinfo(tile->directededge(i));
      const auto& shp = inf.shape();
      const auto& l2 = TileHierarchy::levels().back().tiles;
      auto intersected = l2.Intersect(shp);
      if (intersected.size() > 20) {
        LOG_ERROR(encode(shp));
      }
```

and commenting out the other printing at the bottom of the loop and it printed out some very interesting findings (which ill turn into unit tests, FYI we have unit tests for this but its hard to test every permutation when you are talking about numerical precision):

```
2021/07/09 14:50:12.516705 [INFO] Enumerating edges...
2021/07/09 14:50:32.544482 [INFO] Exporting 711196788 edges
2021/07/09 14:50:32.545076 [INFO] 0%
2021/07/09 14:50:38.223657 [INFO] 1%
2021/07/09 14:50:42.220460 [INFO] 2%
2021/07/09 14:50:46.404522 [INFO] 3%
2021/07/09 14:50:50.169194 [INFO] 4%
2021/07/09 14:50:53.252195 [INFO] 5%
2021/07/09 14:50:56.790646 [INFO] 6%
2021/07/09 14:50:58.861963 [ERROR] {wvueB{knms@GdU
2021/07/09 14:51:00.423221 [INFO] 7%
2021/07/09 14:51:01.734418 [ERROR] }~etdBcnwfk@dE}YnS_oA|OwaAjAwIGudBEuD_Ey^wHwk@]mVxFss@eTu}G_Iuk@}@mZt@mW\sNcLueAN{PvXseBrFq\
2021/07/09 14:51:03.647848 [INFO] 8%
2021/07/09 14:51:07.071192 [INFO] 9%
2021/07/09 14:51:10.150014 [INFO] 10%
2021/07/09 14:51:10.619466 [ERROR] {d{gbBwdpsZ]s~AG}bAFqhAd@goA\_pAUq`B
2021/07/09 14:51:13.109038 [INFO] 11%
2021/07/09 14:51:16.111172 [INFO] 12%
2021/07/09 14:51:17.417675 [ERROR] eibz`Bc_ysVFse@
2021/07/09 14:51:19.068295 [INFO] 13%
2021/07/09 14:51:19.142529 [ERROR] uhbz`BziwcnEUgyl@
2021/07/09 14:51:20.436774 [ERROR] m{~s`B_qbfOu@rDiBxDqCbK]tH?dOd@v\t@x`@l@tZe@dU?bc@Lxh@NrTt@h[\nWN`WmAbTiB`V]vOyClTuDzT
2021/07/09 14:51:22.301245 [INFO] 14%
2021/07/09 14:51:22.842044 [ERROR] _qwg`Bkf}qQgN}g@kAkLiCs`@gCm_@_@wHEcHr@qKbBwK`CqM
2021/07/09 14:51:25.393081 [INFO] 15%
2021/07/09 14:51:28.683880 [INFO] 16%
2021/07/09 14:51:31.587902 [ERROR] cqsa~Aym}kSbBqH|c@ikBd@oH?aEm@eDaC}FwCiG}D}FkGeIyWeVwCgCcAqB}@uD{@gMe@wT_@_UN{I
2021/07/09 14:51:31.865648 [INFO] 17%
2021/07/09 14:51:35.368050 [INFO] 18%
2021/07/09 14:51:38.897027 [INFO] 19%
2021/07/09 14:51:42.062385 [INFO] 20%
2021/07/09 14:51:45.834290 [INFO] 21%
2021/07/09 14:51:49.293962 [INFO] 22%
2021/07/09 14:51:50.489606 [ERROR] swk{yAm{vp\fIiKtJ}[bU}_AbW{{AdEyXl@cNGagAqBut@wD_t@{F}{@wCmy@aCca@mEyYcFyQ
2021/07/09 14:51:51.813468 [ERROR] o_}tyAfwu_iC{FqaDkQgaBiCgaBlEed@rLiWhu@cp@pHk`@t@ce@kLumAav@skDkQ?{P|_@eTlgAaShXuThBipA_s@ag@mJw\aRyRia@uEoSOkqDcKi_Bz_@y_AeE}Tst@qq@kFaSm@mJnNia@L}i@{Jsd@{Vyk@cj@_J_JwNk[wfEyV{k@afAijAcL_^iC_Tf@ce@dYaaEqB}_@
2021/07/09 14:51:53.208868 [INFO] 23%
2021/07/09 14:51:56.824627 [ERROR] wqkdxA_cbeeAG|E
2021/07/09 14:51:56.877365 [INFO] 24%
2021/07/09 14:51:56.951873 [ERROR] q_lgxAkorvq@ud@|}@}@lEEfCLpCd@`CdKnNrPhW|~@|tAv\zg@
2021/07/09 14:52:01.241573 [INFO] 25%
2021/07/09 14:52:05.674697 [INFO] 26%
2021/07/09 14:52:09.482395 [INFO] 27%
2021/07/09 14:52:13.353483 [INFO] 28%
2021/07/09 14:52:17.812729 [INFO] 29%
2021/07/09 14:52:22.010991 [INFO] 30%
2021/07/09 14:52:26.050377 [INFO] 31%
2021/07/09 14:52:29.522488 [ERROR] qbjvpAjpjlcDwDcG{AuEUuEFwMGya@O{~@e@orA]mT
2021/07/09 14:52:30.087014 [INFO] 32%
2021/07/09 14:52:33.981525 [INFO] 33%
2021/07/09 14:52:38.267821 [INFO] 34%
2021/07/09 14:52:42.347034 [INFO] 35%
2021/07/09 14:52:46.080940 [INFO] 36%
2021/07/09 14:52:49.897311 [INFO] 37%
2021/07/09 14:52:52.437690 [ERROR] ehtwkAznrdgELxhB^|@z@z@~eBkAnDm@`CM
2021/07/09 14:52:53.872265 [INFO] 38%
2021/07/09 14:52:57.414984 [INFO] 39%
2021/07/09 14:52:58.579972 [ERROR] uvwijArilewCiCaHmEqGsLl@yLz@}IiBiMOaSfDsPbFkFjBmEtEkB~HFxzE
2021/07/09 14:53:01.321809 [INFO] 40%
2021/07/09 14:53:05.153980 [INFO] 41%
2021/07/09 14:53:09.098028 [INFO] 42%
2021/07/09 14:53:13.101198 [INFO] 43%
2021/07/09 14:53:16.869287 [ERROR] c}kweA~|lpNNnAu@`GgDfG{AdEe@fC{@dAkAZsAA}@cAqBeEwHkQuEmHyBsDe@mD
2021/07/09 14:53:17.242358 [INFO] 44%
2021/07/09 14:53:21.358480 [INFO] 45%
2021/07/09 14:53:23.087918 [ERROR] u`sidA`nzecF]m@W{AMkBF_S
2021/07/09 14:53:25.074407 [INFO] 46%
2021/07/09 14:53:28.329311 [INFO] 47%
2021/07/09 14:53:32.011968 [INFO] 48%
2021/07/09 14:53:35.655795 [INFO] 49%
2021/07/09 14:53:39.503417 [INFO] 50%
2021/07/09 14:53:43.024460 [ERROR] {{at_Ajmkj`F]o|@O{_@?O?aRNekCz@eFbBiBbFkAly@}@jL??g`BlJen@r@m@d~@cz@
```

Those are the encoded polylines, and as you may have guessed are pretty simplistic little shapes. When you plot them a pattern emerges:

![image](https://user-images.githubusercontent.com/697548/125097555-40b82380-e0a4-11eb-9153-9568d6b6111f.png)
![image](https://user-images.githubusercontent.com/697548/125097576-46156e00-e0a4-11eb-8bea-a51e254847d4.png)
![image](https://user-images.githubusercontent.com/697548/125097598-4b72b880-e0a4-11eb-9255-f9ef86a40edb.png)


They all have a section that is perfectly horizontal (at least by eye). Anyway ill keep working on this.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,7 @@
    * FIXED: Fixed incorrect dead-end roundabout labels. [#3129](https://github.com/valhalla/valhalla/pull/3129)
    * FIXED: googletest wasn't really updated in #3166 [#3187](https://github.com/valhalla/valhalla/pull/3187)
    * FIXED: avoid_polygons intersected edges as polygons instead of linestrings [#3194]((https://github.com/valhalla/valhalla/pull/3194)
+   * FIXED: when binning horizontal edge shapes using single precision floats (converted from not double precision floats) allowed for the possiblity of marking many many tiles no where near the shape [#3204](https://github.com/valhalla/valhalla/pull/3204) 
 
 * **Enhancement**
    * CHANGED: Refactor base costing options parsing to handle more common stuff in a one place [#3125](https://github.com/valhalla/valhalla/pull/3125)

--- a/src/midgard/tiles.cc
+++ b/src/midgard/tiles.cc
@@ -17,20 +17,20 @@ namespace {
 // at each step it decides to either move in the x or y direction based on which pixels midpoint
 // forms a smaller triangle with the line. to avoid edge cases we allow set_pixel to make the
 // the loop bail if we leave the valid drawing region
-void bresenham_line(float x0,
-                    float y0,
-                    float x1,
-                    float y1,
+void bresenham_line(double x0,
+                    double y0,
+                    double x1,
+                    double y1,
                     const std::function<bool(int32_t, int32_t)>& set_pixel) {
   // this one for sure
   bool outside = set_pixel(std::floor(x0), std::floor(y0));
   // steps in the proper direction and constants for shoelace formula
-  float sx = x0 < x1 ? 1 : -1, dx = x1 - x0, x = std::floor(x0) + .5f;
-  float sy = y0 < y1 ? 1 : -1, dy = y1 - y0, y = std::floor(y0) + .5f;
+  double sx = x0 < x1 ? 1 : -1, dx = x1 - x0, x = std::floor(x0) + .5f;
+  double sy = y0 < y1 ? 1 : -1, dy = y1 - y0, y = std::floor(y0) + .5f;
   // keep going until we make it to the ending pixel
   while (std::floor(x) != std::floor(x1) || std::floor(y) != std::floor(y1)) {
-    float tx = std::abs(dx * (y - y0) - dy * ((x + sx) - x0));
-    float ty = std::abs(dx * ((y + sy) - y0) - dy * (x - x0));
+    double tx = std::abs(dx * (y - y0) - dy * ((x + sx) - x0));
+    double ty = std::abs(dx * ((y + sy) - y0) - dy * (x - x0));
     // less error moving in the x
     if (tx < ty) {
       x += sx;

--- a/src/midgard/tiles.cc
+++ b/src/midgard/tiles.cc
@@ -32,10 +32,9 @@ void bresenham_line(double x0,
     double tx = std::abs(dx * (y - y0) - dy * ((x + sx) - x0));
     double ty = std::abs(dx * ((y + sy) - y0) - dy * (x - x0));
     // less error moving in the x
-    if (tx < ty) {
+    if (tx < ty || (tx == ty && y0 == y1)) {
       x += sx;
-    }
-    // less error moving in the y
+    } // less error moving in the y
     else {
       y += sy;
     }

--- a/test/tiles.cc
+++ b/test/tiles.cc
@@ -1,6 +1,8 @@
 #include "midgard/tiles.h"
 #include "midgard/aabb2.h"
+#include "midgard/encoded.h"
 #include "midgard/pointll.h"
+#include "midgard/polyline2.h"
 #include "midgard/util.h"
 
 #include <random>
@@ -262,7 +264,7 @@ TEST(Tiles, test_intersect_linestring) {
   assert_answer(t, {{1, 2}, {2, 4}}, intersect_t{{0, {6, 7, 12, 13, 19, 20, 25, 26}}});
   assert_answer(t, {{2, 4}, {1, 2}}, intersect_t{{0, {6, 7, 12, 13, 19, 20, 25, 26}}});
 
-  // some real locations on earth (without polar coordinates accounted for)
+  // some real locations on earth
   Tiles<PointLL> ll(AABB2<PointLL>{-180, -90, 180, 90}, .25, 5);
   std::vector<PointLL> shape{{9.5499754, 47.250248}, {9.55031681, 47.2501144}};
   auto intersection = ll.Intersect(shape);
@@ -275,6 +277,37 @@ TEST(Tiles, test_intersect_linestring) {
   for (const auto& i : intersection)
     count += i.second.size();
   EXPECT_LE(count, 2) << "Unexpected number of intersections for this shape";
+
+  std::vector<std::string> shapes{
+      R"({wvueB{knms@GdU)",
+      R"(}~etdBcnwfk@dE}YnS_oA|OwaAjAwIGudBEuD_Ey^wHwk@]mVxFss@eTu}G_Iuk@}@mZt@mW\sNcLueAN{PvXseBrFq\)",
+      R"({d{gbBwdpsZ]s~AG}bAFqhAd@goA\_pAUq`B)",
+      R"(eibz`Bc_ysVFse@)",
+      R"(uhbz`BziwcnEUgyl@)",
+      R"(m{~s`B_qbfOu@rDiBxDqCbK]tH?dOd@v\t@x`@l@tZe@dU?bc@Lxh@NrTt@h[\nWN`WmAbTiB`V]vOyClTuDzT)",
+      R"(_qwg`Bkf}qQgN}g@kAkLiCs`@gCm_@_@wHEcHr@qKbBwK`CqM)",
+      R"(cqsa~Aym}kSbBqH|c@ikBd@oH?aEm@eDaC}FwCiG}D}FkGeIyWeVwCgCcAqB}@uD{@gMe@wT_@_UN{I)",
+      R"(swk{yAm{vp\fIiKtJ}[bU}_AbW{{AdEyXl@cNGagAqBut@wD_t@{F}{@wCmy@aCca@mEyYcFyQ)",
+      R"(o_}tyAfwu_iC{FqaDkQgaBiCgaBlEed@rLiWhu@cp@pHk`@t@ce@kLumAav@skDkQ?{P|_@eTlgAaShXuThBipA_s@ag@mJw\aRyRia@uEoSOkqDcKi_Bz_@y_AeE}Tst@qq@kFaSm@mJnNia@L}i@{Jsd@{Vyk@cj@_J_JwNk[wfEyV{k@afAijAcL_^iC_Tf@ce@dYaaEqB}_@)",
+      R"(wqkdxA_cbeeAG|E)",
+      R"(q_lgxAkorvq@ud@|}@}@lEEfCLpCd@`CdKnNrPhW|~@|tAv\zg@)",
+      R"(qbjvpAjpjlcDwDcG{AuEUuEFwMGya@O{~@e@orA]mT)",
+      R"(ehtwkAznrdgELxhB^|@z@z@~eBkAnDm@`CM)",
+      R"(uvwijArilewCiCaHmEqGsLl@yLz@}IiBiMOaSfDsPbFkFjBmEtEkB~HFxzE)",
+      R"(c}kweA~|lpNNnAu@`GgDfG{AdEe@fC{@dAkAZsAA}@cAqBeEwHkQuEmHyBsDe@mD)",
+      R"(u`sidA`nzecF]m@W{AMkBF_S)",
+      R"({{at_Ajmkj`F]o|@O{_@?O?aRNekCz@eFbBiBbFkAly@}@jL??g`BlJen@r@m@d~@cz@)",
+      R"(syhxt@kqlv~CyByCkD}IyBkBkD]uD]aA{ANkAdB]nDz@dE\~CMjD{AtBiCt@uDfAuEdAwDTsFlBqG~FuE|GeE`KeEpEeF~C_IzCoHfGaHtGeFbGaGlGsGzCqGrAoI\_IbBaHFqHpBsPt@sQWiL]mJQsF`BuEhCgDfCkBdFqGfFcFnLyMxDmJ~DuErE{K`FaHnC]jEm@jCiBfBwCr@aHl@eFrAgDxB]tCNhDLxCkA`EcGdCgDrCwC)",
+      R"(kchiYyi`x~DPkk@)",
+      R"(r`egLc~vakEf@]d@m@n@gCf@iCFiBIgDQwD?{A@?FM~@_@xA]|A]p@O~CgCnDcG^]h@m@hC}@`A{@^m@DkAUkBkAqGOiCHwCpEmUhCiL\m@^m@^]nDkB`CgCjDeEhCyCpAkAn@{ATyBCyBK{AE{@B]Hm@fAiCD{@A}@aBwNo@oHGmJ)",
+      R"(tcyyTunevjAbc@dhAJpg@zB`GKdd@dIzz@q@vH}IvIaBnXrEv]yE`l@oGpv@)",
+      R"(_n_gnAlrqwuDFyLl@_}@VuwA?sf@Wk}At@qzAe@enALk_@t@afAm@emBl@ocDGg~Cd@e~D]wsDUsxATieOkAojGFse@r@snA?_i@)",
+  };
+  for (const auto& shape : shapes) {
+    auto decoded = decode<std::vector<PointLL>>(shape);
+    auto intersected = ll.Intersect(decoded);
+    ASSERT_LE(intersected.size(), 3) << " too many level 2 tiles intersected by " << shape;
+  }
 }
 
 TEST(Tiles, test_random_linestring) {

--- a/test/tiles.cc
+++ b/test/tiles.cc
@@ -2,7 +2,6 @@
 #include "midgard/aabb2.h"
 #include "midgard/encoded.h"
 #include "midgard/pointll.h"
-#include "midgard/polyline2.h"
 #include "midgard/util.h"
 
 #include <random>

--- a/test/tiles.cc
+++ b/test/tiles.cc
@@ -277,6 +277,16 @@ TEST(Tiles, test_intersect_linestring) {
     count += i.second.size();
   EXPECT_LE(count, 2) << "Unexpected number of intersections for this shape";
 
+  // the following short shapes all end up going to rasterization because they are on the boundary of
+  // subdivisions. when the rasterization algorithm was using single precision floats these shapes
+  // would cause a tie breaker scenario in which it would only rasterize the y direction. the thing is
+  // it was the precision that caused the tie breaker in the first place. switching to double fixes
+  // this because if you make the line long enough to cross more than 2 adjacent bins, it gets
+  // resampled along the geodesic, and when you do this resampling and you do it all in double it
+  // properly projects to the curve which then is never along a perfectly horizontal trajectory
+  // except for at the equator but the resampling also spaces the points out such that they are never
+  // far apart enough to ever do the rasterization step, its all just neighbor pixels and we have a
+  // shortcut for those
   std::vector<std::string> shapes{
       R"({wvueB{knms@GdU)",
       R"(}~etdBcnwfk@dE}YnS_oA|OwaAjAwIGudBEuD_Ey^wHwk@]mVxFss@eTu}G_Iuk@}@mZt@mW\sNcLueAN{PvXseBrFq\)",
@@ -305,8 +315,19 @@ TEST(Tiles, test_intersect_linestring) {
   for (const auto& shape : shapes) {
     auto decoded = decode<std::vector<PointLL>>(shape);
     auto intersected = ll.Intersect(decoded);
-    ASSERT_LE(intersected.size(), 3) << " too many level 2 tiles intersected by " << shape;
+    ASSERT_LE(intersected.size(), 3) << "Too many level 2 tiles intersected by " << shape;
   }
+
+  // horizontal over 3 subdivisions (bins) so that its forced to do bresenham rasterization. also,
+  // note that we do this one in the plane to avoid resampling (adding points along the line) which is
+  // done in the spherical algorithm to account for the curvature of the geodescic. we dont want that
+  // to happen because inserting extra points reverts the test case to a simple sequence of
+  // neighboring bins which takes away our ability to test this a regression in which perfectly
+  // horizontal rasterization marked way too many subdivisions
+  Tiles<Point2> planar(AABB2<Point2>{-180, -90, 180, 90}, .25, 5);
+  auto rasterized = planar.Intersect(std::vector<Point2>{{0.04, 0}, {0.11, 0}});
+  ASSERT_EQ(rasterized.size(), 1) << "One tile should be intersected";
+  ASSERT_EQ(rasterized.begin()->second.size(), 3) << "Three subtiles (bins) should be intersected";
 }
 
 TEST(Tiles, test_random_linestring) {

